lib/phoenix_component.ex

```diff
@@ -290,6 +290,139 @@ defmodule Phoenix.Component do
     quote do
       import Phoenix.LiveView
       import Phoenix.LiveView.Helpers
+      import unquote(__MODULE__), only: [attr: 2, attr: 3]
+
+      Module.register_attribute(__MODULE__, :__attrs__, accumulate: true)
+      Module.register_attribute(__MODULE__, :__components_calls__, accumulate: true)
+      Module.put_attribute(__MODULE__, :__components__, %{})
+
+      @on_definition {unquote(__MODULE__), :__on_definition__}
+      @before_compile {unquote(__MODULE__), :__before_compile__}
+    end
+  end
+
+  @doc "Defines an attribute for the component"
+  defmacro attr(name, type, opts \\ []) do
+    quote bind_quoted: [
+            name: name,
+            type: type,
+            opts: opts,
+            line: __CALLER__.line
+          ] do
+      Phoenix.Component.validate_attr!(name, type, opts, line, __ENV__.file)
+      Module.put_attribute(__MODULE__, :__attrs__, %{name: name, type: type, opts: opts, line: line})
```

Comment:
To avoid the amount of injected code, move the call to put_attribute inside `validate_attr!`. Also, since `Phoenix.Component` is imported, we cannot have a function named `validate_attr!`, please rename it to `__attr__!` instead. :)